### ISSUE TITLE
Differences between Octave.app and direct Homebrew install

### ISSUE THREAD

**Issue by rdzman** [octave-app/octave-app#271]
The test suite for my power systems analysis/optimization package, [MATPOWER](https://matpower.org) passes just fine under Octave installed directly via Homebrew _(e.g. v 9.1.0 on macOS 14.x on M2 Max MBP, previously v8.4.0 on macOS 13.x on same machine)._ But when I use recent builds of Octave.app _(such as the recent 8.4.0_RC1)_, a few of the tests that rely on `glpk()` are failing due to minor, though non-trivial, numerical differences in the solutions. FWIW, I don't notice this issue with older builds _(e.g. 6.3.0-beta1)._

Do you have ideas what would cause this? Differences in some math library between the OS of the Octave.app build environment and mine? Anything we could try to avoid the issue and give me something closer to what Homebrew installs by default?

**Comment by apjanke** [comment]
> Do you have ideas what would cause this? Differences in some math library between the OS of the Octave.app build environment and mine?

This is entirely possible. Two main differences I can think of: version differences in the math libraries, and compilation differences.

Version differences: When Octave.app is built, it builds and includes all the dependencies of octave inside the build, including the math libraries, using the latest versions in the Homebrew formulae for them in core Homebrew and in the [octave-app Tap](https://github.com/octave-app/homebrew-octave-app). Those versions are fixed at Octave.app build time, and that's what you get when you install Octave.app, and they're never updated (until you install a later Octave.app release that comes with newer versions of those libraries). When you install Octave with Homebrew (like `brew install octave`), you get the latest versions

Compilation differences: I build Octave.app on my local Macs, while Homebrew builds their bottle packages using GitHub Action runner machines (I think). They may not be running exactly the same CPU generations, which could lead to some differences in how the code is built, in terms of both optimizations and what instructions are used. I currently use an 2017 iMac Pro with Xeon W 10-core CPU for the Intel builds and a 2021 MacBook Pro M1 for the Apple Silicon builds. The Homebrew build process uses conservative architecture choices in the build for portability, but I'm not sure exactly how that works, especially when doing a regular build from source due to a non-default installation location (like Octave.app does) instead of the special bottle-building process. In fact, that difference in non-bottle builds might account for this, because Octave.app may be built targeting a newer CPU generation than the conservative bottle builds, and using newer CPU instructions and optimizations.

Non-math libraries, like the GCC compiler, could also affect results here.

> Anything we could try to avoid the issue and give me something closer to what Homebrew installs by default?

I don't know if I can make Octave.app actually match or track what a current Homebrew build is producing. I can't use their "bottle" builds because those are tied to the installation location. And Octave.app is always going to have fixed versions of the math etc library dependencies in each release.

But we can do stuff to maybe figure out where the difference is coming from, and either correct it if it's an actual issue in the math in the Octave.app builds, or make the test accommodate the difference if it's just a legit "floating point arithmetic is approximate and your results may vary between machines etc." thing.

For examining the version differences between the packages you have installed in your Homebrew, and the packages used by the Octave.app you're running, the Octave.app bundle has lists of the package versions at `Contents/Resources/VERSIONS` and, in newer versions, in more detail at `Contents/Resources/octapp-meta`, inside the app bundle. Have a look at those. I don't have a tool to do automatic diffing, but it wouldn't be hard to put one together. To do a diff manually, run `brew list --versions` and save that to a file, and diff it against the `octapp-meta/brew-list-versions.txt` file (I think) from the Octave.app app bundle.

To test for formula and build differences, you can use the [octave-app Tap](https://github.com/octave-app/homebrew-octave-app) to install the special Octave.app ("octapp") versions of the formulae I use to build Octave.app, but in to your regular Homebrew installation. Do this:

* `brew tap octave-app/octave-app`
* `brew install octave-octapp@8.4.0`

That will install the custom Octave.app versions of Octave and some of its dependencies in to your own local Homebrew installation. It's a side-by-side installation that won't conflict with any core Homebrew formulae (including `octave`), so you don't have to remove anything or (hopefully) worry about side effects. The build will probably take a couple hours; sorry. Once you have that, you should be able to run `octave-octapp@8.4.0` to launch the octapp version of Octave from your Homebrew. Run the tests there and see how they go. That'll give you some info. If the test results are the same as the vanilla core Homebrew, that indicates it may be a dependency-version difference. If the test results match your Octave.app results, then it may be a formula or build-process definition.

You can also try doing local builds of the math libraries in your local Homebrew to see if it's a compilation/build difference between the Homebrew binary bottle distributions and local builds like Octave.app does (which may use newer CPU architecture features). To do that, `brew reinstall --build-from-source <formula>`. You'd want to do Octave, openblas, and whatever other math libraries you think may be involved. Again, a slow build, and this will affect the main packages you have installed in your Homebrew and any other apps that use them. So may not want to do it on your main machine.

I can also do these things myself if I find some time :) and you describe the exact test result differences you're seeing. This is something I'm interested in understanding.

> FWIW, I don't notice this issue with older builds (e.g. 6.3.0-beta1).

Could be a few things: maybe the dependency versions were closer matched due to build cadence and when you ran the tests, maybe I was using build boxes with CPUs closer to the Homebrew build boxes (and at the time, I believe core Homebrew was running their build farm on a different service), maybe it's an Apple Silicon thing, maybe I introduced some relevant formula changes between then and now, or the core Homebrew build process diverged (like whether they use GCC or clang/Xcode to build the math libraries; Octave.app forces GCC for reasons that I don't know or remember the origins of).

**Comment by rdzman** [comment]
Wow. Thank you for all of this very helpful information.

In the end, I ran into what appeared to be the same numerical issue on some other Octave builds (Linux GitHub runners), so I decided to focus on finding a way to make my tests a bit more numerically robust and to avoid the issue completely. As I'm sure you know, It's not unusual with some numerical algorithms that small differences can result in very different results. Fortunately, I was able to find a solution that seems to work consistently across builds, including the latest Octave.app builds.

Thanks again for your valuable work on this project.
